lib/SILOptimizer/IPO/CrossModuleOptimization.cpp

```diff
@@ -280,9 +310,21 @@ bool CrossModuleOptimization::canSerializeInstruction(SILInstruction *inst,
     // public functions, because that can increase code size. E.g. if the
     // function is completely inlined afterwards.
     // Also, when emitting TBD files, we cannot introduce a new public symbol.
-    if ((conservative || M.getOptions().emitTBD) &&
-        !hasPublicOrPackageVisibility(callee->getLinkage(), M.getOptions().EnableSerializePackage)) {
-      return false;
+    if (conservative || M.getOptions().emitTBD) {
+      // In package-cmo, serialize this inst if its referenced function has
+      // package+ linkage, or is a definition with shared linkage.
+      // For example, `public func foo() { print("") }` is a function with
+      // public linkage which contains `print`, which is a definition with
+      // shared linkage, thus should be serialized.
+      if (M.getOptions().EnableSerializePackage) {
+        if (!(hasPublicOrPackageVisibility(callee->getLinkage(),
+                                           /*includePackage*/ true) ||
+              (hasSharedVisibility(callee->getLinkage()) &&
```

Comment:
We need to know that the `shared` `callee` can also safely be serialized.

For now, you can add `callee->isSerialized()` as a conservative approximation.

Reply:
The callee (SIL definition for `print` in the example in the comment) is not necessarily serialized by this point; it gets serialized later, thus having that check would prevent a function referencing it from being serialized. If we must enforce the check, I suppose we could reorder the list of SIL functions in the module in the order of SILLinkage (shared, then package/public), but not sure how necessary that would be.